src/twisted/internet/abstract.py

```diff
@@ -242,13 +242,17 @@ def doWrite(self):
 
         @see: L{twisted.internet.interfaces.IWriteDescriptor.doWrite}.
         """
-        if len(self.dataBuffer) - self.offset < self.SEND_LIMIT:
+        remaining = len(self.dataBuffer) - self.offset
+        if remaining < self.SEND_LIMIT:
             # If there is currently less than SEND_LIMIT bytes left to send
             # in the string, extend it with the array data.
-            self.dataBuffer = _concatenate(
-                self.dataBuffer, self.offset, self._tempDataBuffer
-            )
-            self.offset = 0
+            if remaining:
```

Comment:
I think I know what is going on here ... but I had to read this code a couple of times.

Can we have comments for `if remaining: else` branches ?

--------

Can we update the declaration of `offset` on line 174 to document what `self.offset` is expected to do and why it is public API.

--------

I have checked the definitions of twisted/internet/interfaces.py and I can't find a mention for `offset`

Maybe, it should be private API

----------

Also, what is `SEND_LIMIT` or `self.dataBuffer`  ?

Maybe have an explicit comparison and a comment

```suggestion
            if remaining > 0:
                # We still have data not sent from the previous buffer.
                # Add the new data to the existing buffer.
            
```